### DISCUSSION TITLE
Noisy Quantum Circuit simulation - locally purified structure (LPDO)

### DISCUSSION THREAD

**Discussion by EmilianoG-byte** [tencent-quantum-lab/tensorcircuit#166] [Q&A]
Hello :)

I am interested in simulating noisy quantum circuits. In particular, I would like to follow the locally purified density operator (LPDO) structure from this [paper](https://arxiv.org/abs/1412.5746).

What would be the best way to implement this within `TensorCircuit`? As I understand, noisy circuits are currently simulated in `TensorCircuit` either through MonteCarlo trajectories with MPS, or with the full density matrix as MPO. The approach I would like to explore is rather a third option, using only the purification of the density matrix.

Thank you in advance for your help!

**Comment by refraction-ray** [comment]
A very nice question, and indeed to have LPDO simulation engine in TensorCircuit would be super nice. Currently, if you want to simulate something like this, you have to manipulate the lower level API of TensorCircuit and TensorNetwork to explicitly implement the algorithm, though I am confident that all primitives are here and no main obstacle avoiding the implementation in TensorCircuit. A reference on the lower API usage and the implementation could be found https://github.com/tencent-quantum-lab/tensorcircuit/blob/b00279638c93044748c4ed74560a92ef180b6ff0/tensorcircuit/densitymatrix.py#L331. If you are willing to implement this using TensorCircuit, I am more than welcome, and we might further discuss the plan to merge the code into TensorCircuit codebase if you like.
